core/bin/dust_api.rs

```diff
@@ -1641,6 +1641,250 @@ async fn data_sources_delete(
+            Some(e),
+        ),
+        Ok(dbs) => (
+            StatusCode::OK,
+            Json(APIResponse {
+                error: None,
+                response: Some(json!({
+                    "databases": dbs
+                })),
+            }),
+        ),
+    }
+}
+
+#[derive(serde::Deserialize)]
+struct DatabasesTablesUpsertPayload {
+    table_id: String,
+    name: String,
+    description: String,
+}
+
+async fn databases_tables_upsert(
+    extract::Path((project_id, data_source_id, database_id)): extract::Path<(i64, String, String)>,
+    extract::Json(payload): extract::Json<DatabasesTablesUpsertPayload>,
+    extract::Extension(state): extract::Extension<Arc<APIState>>,
+) -> (StatusCode, Json<APIResponse>) {
+    let project = project::Project::new_from_id(project_id);
+
+    match state
+        .store
+        .upsert_database_table(
+            &project,
+            &data_source_id,
+            &database_id,
+            &payload.table_id,
+            &payload.name,
+            &payload.description,
+        )
+        .await
+    {
+        Err(e) => error_response(
+            StatusCode::INTERNAL_SERVER_ERROR,
+            "internal_server_error",
+            "Failed to upsert database table",
+            Some(e),
+        ),
+        Ok(table) => (
+            StatusCode::OK,
+            Json(APIResponse {
+                error: None,
+                response: Some(json!({
+                    "table": table
+                })),
+            }),
+        ),
+    }
+}
+
+async fn databases_tables_retrieve(
+    extract::Path((project_id, data_source_id, database_id, table_id)): extract::Path<(
+        i64,
+        String,
+        String,
+        String,
+    )>,
+    extract::Extension(state): extract::Extension<Arc<APIState>>,
+) -> (StatusCode, Json<APIResponse>) {
+    let project = project::Project::new_from_id(project_id);
+
+    match state
+        .store
+        .load_database_table(&project, &data_source_id, &database_id, &table_id)
+        .await
+    {
+        Err(e) => error_response(
+            StatusCode::INTERNAL_SERVER_ERROR,
+            "internal_server_error",
+            "Failed to retrieve database table",
+            Some(e),
+        ),
+        Ok(table) => match table {
+            None => error_response(
+                StatusCode::NOT_FOUND,
+                "database_table_not_found",
+                &format!("No database table found for id `{}`", table_id),
+                None,
+            ),
+            Some(table) => (
+                StatusCode::OK,
+                Json(APIResponse {
+                    error: None,
+                    response: Some(json!({
+                        "table": table
+                    })),
+                }),
+            ),
+        },
+    }
+}
+
+#[derive(serde::Deserialize)]
+struct DatabasesTablesListQuery {
+    offset: usize,
+    limit: usize,
+}
+
+async fn databases_tables_list(
```

Comment:
I think there is no world in which we will not return all the tables? I'd remove offset/limit all the way down and up

Reply:
You're right, the offet/limit part is maybe not super helpful.

> all the way down and up

Reply:
Do you mean at HTTP API level + at store level ?

Reply:
I've removed in from the HTTP api (still optionally available in `store`)